File: src/lib/audit/token-audit.ts
import {
	existsSync,
	readdirSync,
	readFileSync,
	type Stats,
	statSync,
	writeFileSync,
} from "node:fs";
import { join } from "node:path";
import { getEncoding } from "js-tiktoken";

const DEFAULT_COST_PER_1M_TOKENS = 3.0;

let enc: ReturnType<typeof getEncoding> | null = null;
function getEnc() {
	if (!enc) {
		try {
			enc = getEncoding("cl100k_base");
		} catch {
			/* unavailable */
		}
	}
	return enc;
}

export function estimateTokens(
	text: string,
	useTiktoken: boolean = true,
): [number, string] {
	if (useTiktoken) {
		const encoder = getEnc();
		if (encoder) {
			try {
				return [encoder.encode(text).length, "[exact]"];
			} catch {
				/* fall through */
			}
		}
	}
	return [Math.floor(text.length / 4), "[estimated]"];
}

function walkFiles(dir: string): string[] {
	const out: string[] = [];
	let entries: string[];
	try {
		entries = readdirSync(dir);
	} catch {
		return out;
	}
	for (const entry of entries) {
		const fp = join(dir, entry);
		let st: Stats;
		try {
			st = statSync(fp);
		} catch {
			continue;
		}
		if (st.isDirectory()) {
			out.push(...walkFiles(fp));
		} else if (entry.endsWith(".md") || entry.endsWith(".sh")) {
			out.push(fp);
		}
	}
	return out;
}

export function scanFiles(skillDir: string): Record<string, number> {
	const results: Record<string, number> = {};
	for (const f of walkFiles(skillDir).sort()) {
		const rel = f
			.slice(skillDir.length)
			.replace(/^[\\/]/, "")
			.replace(/\\/g, "/");
		const text = readFileSync(f, "utf-8");
		const [tokens] = estimateTokens(text);
		results[rel] = tokens;
	}
	return results;
}

export function parseAlwaysRead(gatewayPath: string): string[] {
	let content: string;
	try {
		content = readFileSync(gatewayPath, "utf-8");
	} catch {
		return [];
	}
	const files: string[] = [];
	let inSection = false;
	for (const line of content.split("\n")) {
		if (/##\s+Always Read/i.test(line)) {
			inSection = true;
			continue;
		}
		if (inSection && line.startsWith("##")) break;
		if (inSection) {
			const m = line.match(/`?rules\/([^`\s]+)`?/);
			if (m) files.push(`rules/${m[1]}`);
		}
	}
	return files;
}

export function parseCommonTasks(
	gatewayPath: string,
): Record<string, string[]> {
	let content: string;
	try {
		content = readFileSync(gatewayPath, "utf-8");
	} catch {
		return {};
	}
	const tasks: Record<string, string[]> = {};
	const sectionMatch = content.match(/##\s+Common Tasks.*?(\n##\s|Z)/is);
	if (!sectionMatch) return tasks;
	const lines = sectionMatch[0]
		.split("\n")
		.filter((l) => l.trim().startsWith("|"));
	for (const line of lines.slice(2)) {
		const cells = line
			.split("|")
			.map((c) => c.trim())
			.filter((c) => c);
		if (
			cells.length >= 3 &&
			cells[0].toLowerCase() !== "task" &&
			!cells[0].startsWith("<!--")
		) {
			const taskName = cells[0]
				.replace(/[^\w\s]/g, "")
				.trim()
				.toLowerCase()
				.replace(/\s+/g, "_");
			const refs = cells[1].match(/`([^`]+\.(?:md|mdc|sh|py))`/g) || [];
			tasks[taskName] = refs.map((r) => r.replace(/`/g, ""));
		}
	}
	return tasks;
}

export interface AuditSkillResult {
	skill_name: string;
	method: string;
	single_file_tokens: Record<string, number>;
	naive_all_tokens: number;
	l0_tokens: number;
	gateway_tokens: number;
	per_task_tokens: Record<string, number>;
	session_5rounds: {
		naive_total_tokens: number;
		crp_total_tokens: number;
		savings_percent: number;
	};
	cost_per_1m_tokens: number;
	cost_usd: { naive: number; crp: number; savings: number };
}

export function auditSkill(
	skillDir: string,
	skillName: string,
	useTiktoken: boolean = true,
	rounds: number = 5,
	scenario?: string[],
): AuditSkillResult {
	const gateway = join(skillDir, "SKILL.md");
	if (!existsSync(gateway)) throw new Error(`${gateway} not found`);

	const fileTokens = scanFiles(skillDir);
	const naiveTotal = Object.values(fileTokens).reduce((a, b) => a + b, 0);

	const l0Files = parseAlwaysRead(gateway);
	const l0Tokens = l0Files.reduce((sum, f) => sum + (fileTokens[f] || 0), 0);

	const gatewayContent = readFileSync(gateway, "utf-8");
	const [gatewayTokens, methodLabel] = estimateTokens(
		gatewayContent,
		useTiktoken,
	);

	const tasks = parseCommonTasks(gateway);
	const perTask: Record<string, number> = {};
	for (const [taskName, refs] of Object.entries(tasks)) {
		const taskSpecific = refs.reduce((sum, f) => sum + (fileTokens[f] || 0), 0);
		perTask[taskName] = l0Tokens + gatewayTokens + taskSpecific;
	}
	perTask.other_unlisted = l0Tokens + gatewayTokens;

	const defaultPattern = ["fix_bug", "add_feature", "other_unlisted"];
	const roundTasks =
		scenario ||
		Array.from(
			{ length: rounds },
			(_, i) => defaultPattern[i % defaultPattern.length],
		);

	const sessionNaive = naiveTotal * rounds;
	const sessionCrp = roundTasks.reduce(
		(sum, t) => sum + (perTask[t] || perTask.other_unlisted),
		0,
	);

	const costNaive = (sessionNaive / 1_000_000) * DEFAULT_COST_PER_1M_TOKENS;
	const costCrp = (sessionCrp / 1_000_000) * DEFAULT_COST_PER_1M_TOKENS;

	return {
		skill_name: skillName,
		method: methodLabel,
		single_file_tokens: fileTokens,
		naive_all_tokens: naiveTotal,
		l0_tokens: l0Tokens,
		gateway_tokens: gatewayTokens,
		per_task_tokens: perTask,
		session_5rounds: {
			naive_total_tokens: sessionNaive,
			crp_total_tokens: sessionCrp,
			savings_percent: sessionNaive
				? Math.round((1 - sessionCrp / sessionNaive) * 1000) / 10
				: 0,
		},
		cost_per_1m_tokens: DEFAULT_COST_PER_1M_TOKENS,
		cost_usd: {
			naive: Math.round(costNaive * 10000) / 10000,
			crp: Math.round(costCrp * 10000) / 10000,
			savings: Math.round((costNaive - costCrp) * 10000) / 10000,
		},
	};
}

export function runAudit(
	skillName: string | null,
	skills: Array<{ name: string }>,
	useTiktoken: boolean,
	report: boolean = false,
	rounds: number = 5,
	scenario?: string[],
): number {
	const skillsDir = ".claude/skills";

	if (skills.length === 1 || skillName) {
		const name = skillName || skills[0]?.name;
		if (!name) {
			console.log("ERROR: No skill specified");
			return 1;
		}
		const skillDir = join(skillsDir, name);
		const result = auditSkill(skillDir, name, useTiktoken, rounds, scenario);
		printSingleAudit(result, rounds);
		if (report) {
			writeFileSync(
				"benchmark-report.json",
				`${JSON.stringify(result, null, 2)}\n`,
				"utf-8",
			);
			console.log("\nReport written to benchmark-report.json");
		}
		return 0;
	}

	const allResults: Record<string, AuditSkillResult> = {};
	let totalNaive = 0;
	let totalCrp = 0;
	for (const skill of skills) {
		const skillDir = join(skillsDir, skill.name);
		if (!existsSync(skillDir)) {
			console.log(`WARNING: Skill directory not found: ${skillDir}`);
			continue;
		}
		const result = auditSkill(
			skillDir,
			skill.name,
			useTiktoken,
			rounds,
			scenario,
		);
		allResults[skill.name] = result;
		totalNaive += result.session_5rounds.naive_total_tokens;
		totalCrp += result.session_5rounds.crp_total_tokens;
	}

	const combinedSavings = totalNaive
		? Math.round((1 - totalCrp / totalNaive) * 1000) / 10
		: 0;
	console.log(
		`\n== Multi-Skill Token Audit (${Object.keys(allResults).length} skills) ==\n`,
	);
	console.log(`Naive total:    ${totalNaive.toLocaleString()} tokens`);
	console.log(`CRP total:      ${totalCrp.toLocaleString()} tokens`);
	console.log(`Combined savings: ${combinedSavings}%`);

	if (report) {
		writeFileSync(
			"benchmark-report.json",
			`${JSON.stringify(
				{
					combined_analysis: {
						total_skills: Object.keys(allResults).length,
						naive_total_tokens: totalNaive,
						crp_total_tokens: totalCrp,
						savings_percent: combinedSavings,
					},
					individual_skills: allResults,
				},
				null,
				2,
			)}\n`,
			"utf-8",
		);
		console.log("\nReport written to benchmark-report.json");
	}
	return 0;
}

function printSingleAudit(result: AuditSkillResult, rounds: number): void {
	console.log(`\n== Token Audit: ${result.skill_name} ==\n`);
	console.log(`Estimation method: ${result.method}`);
	console.log(
		`Naive load (all files):     ${result.naive_all_tokens.toLocaleString()} tokens`,
	);
	console.log(
		`L0 (Always Read):           ${result.l0_tokens.toLocaleString()} tokens`,
	);
	console.log(
		`L2 (Gateway/SKILL.md):      ${result.gateway_tokens.toLocaleString()} tokens`,
	);
	console.log("\nPer-task CRP load:");
	for (const [task, tokens] of Object.entries(result.per_task_tokens)) {
		console.log(
			`  ${task.padEnd(20)} ${tokens.toLocaleString().padStart(8)} tokens`,
		);
	}
	const sr = result.session_5rounds;
	console.log(`\n${rounds}-round session simulation:`);
	console.log(
		`  Naive total:  ${sr.naive_total_tokens.toLocaleString()} tokens`,
	);
	console.log(`  CRP total:    ${sr.crp_total_tokens.toLocaleString()} tokens`);
	console.log(`  Savings:      ${sr.savings_percent}%`);
	const c = result.cost_usd;
	console.log(
		`\nEstimated input cost (Claude Sonnet 4.6, $${DEFAULT_COST_PER_1M_TOKENS}/1M tokens):`,
	);
	console.log(`  Naive: $${c.naive.toFixed(4)}`);
	console.log(`  CRP:   $${c.crp.toFixed(4)}`);
	console.log(`  Saved: $${c.savings.toFixed(4)}`);
}

File: scripts/token-audit.py
#!/usr/bin/env python3
"""token-audit.py — Estimate and compare token costs for CRP vs naive loading.

Scans all markdown files in the project and estimates token consumption
under two regimes:
    1. Naive: load every .md file on every task
    2. CRP: load only L0 (Always Read) + L1 (task-specific) files

Token estimation uses tiktoken when available and configured.
Install `tiktoken` for more accurate counts: pip install tiktoken

Usage:
    python scripts/token-audit.py [--skill <name>] [--report]

    --skill   Skill directory to audit (default: auto-detect)
    --report  Write JSON report to benchmark-report.json
"""

from __future__ import annotations

import argparse
import json
import re
import sys
from pathlib import Path

from crp_manifest import load_manifest

DEFAULT_COST_PER_1M_TOKENS = 3.0


def estimate_tokens(text: str, use_tiktoken: bool = True) -> tuple[int, str]:
    """Estimate token count. Returns (count, method_label).

    method_label is '[exact]' for tiktoken, '[estimated]' for heuristic.
    """
    if use_tiktoken:
        try:
            import tiktoken

            enc = tiktoken.get_encoding("cl100k_base")
            return len(enc.encode(text)), "[exact]"
        except ImportError:
            pass
    return len(text) // 4, "[estimated]"


def scan_files(skill_dir: Path) -> dict[str, int]:
    """Return mapping of relative path → token count for all .md and .sh files."""
    results = {}
    for pattern in ("*.md", "*.sh"):
        for f in sorted(skill_dir.rglob(pattern)):
            rel = f.relative_to(skill_dir).as_posix()
            text = f.read_text(encoding="utf-8")
            results[rel], _ = estimate_tokens(text)
    return results


def parse_always_read(gateway_path: Path) -> list[str]:
    """Extract Always Read file list from gateway.md."""
    content = gateway_path.read_text(encoding="utf-8")
    files = []
    in_section = False
    for line in content.splitlines():
        if re.search(r"##\s+Always Read", line, re.IGNORECASE):
            in_section = True
            continue
        if in_section and line.startswith("##"):
            break
        if in_section:
            m = re.search(r"`?rules/([^`\s]+)`?", line)
            if m:
                files.append(f"rules/{m.group(1)}")
    return files


def parse_common_tasks(gateway_path: Path) -> dict[str, list[str]]:
    """Extract per-task file loads from Common Tasks table."""
    content = gateway_path.read_text(encoding="utf-8")
    tasks = {}
    section_match = re.search(
        r"##\s+Common Tasks.*?(\n##\s|\Z)", content, re.DOTALL | re.IGNORECASE
    )
    if not section_match:
        return tasks

    section = section_match.group(0)
    lines = [ln for ln in section.splitlines() if ln.strip().startswith("|")]
    for line in lines[2:]:
        cells = [c.strip() for c in line.split("|")]
        cells = [c for c in cells if c]
        if len(cells) >= 3 and cells[0].lower() not in ("task", "<!-- fill:"):
            raw_name = cells[0]
            task_name = re.sub(r"[^\w\s]", "", raw_name).strip().lower().replace(" ", "_")
            reads = cells[1]
            refs = re.findall(r"`([^`]+\.(?:md|mdc|sh|py))`", reads)
            tasks[task_name] = refs
    return tasks


def audit_skill(skill_dir: Path, skill_name: str, use_tiktoken: bool = True) -> dict:
    gateway = skill_dir / "SKILL.md"
    if not gateway.exists():
        print(f"ERROR: {gateway} not found")
        sys.exit(1)

    file_tokens = scan_files(skill_dir)
    naive_total = sum(file_tokens.values())

    l0_files = parse_always_read(gateway)
    l0_tokens = sum(file_tokens.get(f, 0) for f in l0_files)

    gateway_tokens, method_label = estimate_tokens(gateway.read_text(encoding="utf-8"), use_tiktoken)

    tasks = parse_common_tasks(gateway)
    per_task = {}
    for task_name, refs in tasks.items():
        task_specific = sum(file_tokens.get(f, 0) for f in refs)
        per_task[task_name] = l0_tokens + gateway_tokens + task_specific

    per_task["other_unlisted"] = l0_tokens + gateway_tokens

    round_tasks = ["fix_bug", "add_feature", "other_unlisted", "fix_bug", "add_feature"]
    session_naive = naive_total * 5
    session_crp = sum(per_task.get(t, per_task["other_unlisted"]) for t in round_tasks)

    cost_naive = (session_naive / 1_000_000) * DEFAULT_COST_PER_1M_TOKENS
    cost_crp = (session_crp / 1_000_000) * DEFAULT_COST_PER_1M_TOKENS

    return {
        "skill_name": skill_name,
        "method": method_label,
        "single_file_tokens": file_tokens,
        "naive_all_tokens": naive_total,
        "l0_tokens": l0_tokens,
        "gateway_tokens": gateway_tokens,
        "per_task_tokens": per_task,
        "session_5rounds": {
            "naive_total_tokens": session_naive,
            "crp_total_tokens": session_crp,
            "savings_percent": round((1 - session_crp / session_naive) * 100, 1) if session_naive else 0,
        },
        "cost_per_1m_tokens": DEFAULT_COST_PER_1M_TOKENS,
        "cost_usd": {
            "naive": round(cost_naive, 4),
            "crp": round(cost_crp, 4),
            "savings": round(cost_naive - cost_crp, 4),
        },
    }


def get_skills_to_audit(skill_name: str | None = None) -> list[tuple[str, Path]]:
    """Determine which skills to audit. Returns list of (skill_name, skill_dir)."""
    manifest = load_manifest(Path("crp.yaml"))
    is_v21 = bool(manifest and manifest.get("version") == "2.1")

    if skill_name:
        if any(c in skill_name for c in "./\\"):
            print(f"ERROR: Invalid skill name: {skill_name!r}")
            sys.exit(1)
        skill_dir = Path(f".claude/skills/{skill_name}")
        if not skill_dir.exists():
            skill_dir = Path(f"templates/{skill_name}")
        return [(skill_name, skill_dir)]

    if is_v21:
        skills = manifest.get("skills", [])
        result = []
        for skill in skills:
            name = skill["name"]
            skill_dir = Path(f".claude/skills/{name}")
            if skill_dir.exists():
                result.append((name, skill_dir))
        return result

    # v2.0 fallback: auto-detect single skill
    skills_dir = Path(".claude/skills")
    if skills_dir.exists():
        subdirs = [d for d in skills_dir.iterdir() if d.is_dir() and d.name != "shared"]
        if len(subdirs) == 1:
            return [(subdirs[0].name, subdirs[0])]
        elif len(subdirs) > 1:
            print("Multiple skills found. Use --skill to specify one:")
            for d in subdirs:
                print(f"  --skill {d.name}")
            sys.exit(1)

    template_skill = Path("templates/skill")
    if template_skill.exists():
        return [("skill", template_skill)]

    print("ERROR: Could not auto-detect skill. Use --skill.")
    sys.exit(1)


def run_audit(skill_name: str | None = None, report: bool = False) -> int:
    """Core audit logic. Called by main() and by crp-setup.py."""
    manifest = load_manifest(Path("crp.yaml"))
    use_tiktoken = True
    if manifest and manifest.get("audit", {}).get("use_tiktoken") is False:
        use_tiktoken = False

    skills = get_skills_to_audit(skill_name)

    if len(skills) == 1:
        # Single skill audit (v2.0 style output)
        name, skill_dir = skills[0]
        if not skill_dir.exists():
            print(f"ERROR: Skill directory not found: {skill_dir}")
            return 1

        result = audit_skill(skill_dir, name, use_tiktoken)

        print(f"\n== Token Audit: {name} ==\n")
        print(f"Estimation method: {result['method']}")
        print(f"Naive load (all files):     {result['naive_all_tokens']:,} tokens")
        print(f"L0 (Always Read):           {result['l0_tokens']:,} tokens")
        print(f"L2 (Gateway/SKILL.md):      {result['gateway_tokens']:,} tokens")
        print("\nPer-task CRP load:")
        for task, tokens in result["per_task_tokens"].items():
            print(f"  {task:20s} {tokens:>8,} tokens")

        sr = result["session_5rounds"]
        print(f"\n5-round session simulation:")
        print(f"  Naive total:  {sr['naive_total_tokens']:,} tokens")
        print(f"  CRP total:    {sr['crp_total_tokens']:,} tokens")
        print(f"  Savings:      {sr['savings_percent']}%")

        c = result["cost_usd"]
        cost_label = DEFAULT_COST_PER_1M_TOKENS
        print(f"\nEstimated input cost (Claude Sonnet 4.6, ${cost_label}/1M tokens):")
        print(f"  Naive: ${c['naive']:.4f}")
        print(f"  CRP:   ${c['crp']:.4f}")
        print(f"  Saved: ${c['savings']:.4f}")

        if report:
            out_path = Path("benchmark-report.json")
            out_path.write_text(json.dumps(result, indent=2, ensure_ascii=False) + "\n", encoding="utf-8")
            print(f"\nReport written to {out_path}")

        return 0

    # Multi-skill combined audit
    all_results = {}
    total_naive = 0
    total_crp = 0

    for name, skill_dir in skills:
        if not skill_dir.exists():
            print(f"WARNING: Skill directory not found: {skill_dir}")
            continue
        result = audit_skill(skill_dir, name, use_tiktoken)
        all_results[name] = result
        sr = result["session_5rounds"]
        total_naive += sr["naive_total_tokens"]
        total_crp += sr["crp_total_tokens"]

    combined_savings = round((1 - total_crp / total_naive) * 100, 1) if total_naive else 0
    combined = {
        "total_skills": len(all_results),
        "naive_total_tokens": total_naive,
        "crp_total_tokens": total_crp,
        "savings_percent": combined_savings,
        "cost_usd": {
            "naive": round((total_naive / 1_000_000) * DEFAULT_COST_PER_1M_TOKENS, 4),
            "crp": round((total_crp / 1_000_000) * DEFAULT_COST_PER_1M_TOKENS, 4),
            "savings": round(((total_naive - total_crp) / 1_000_000) * DEFAULT_COST_PER_1M_TOKENS, 4),
        },
    }

    print(f"\n== Multi-Skill Token Audit ({len(all_results)} skills) ==\n")
    print(f"Estimation method: {list(all_results.values())[0]['method'] if all_results else '[estimated]'}")
    print(f"Naive total:    {total_naive:,} tokens")
    print(f"CRP total:      {total_crp:,} tokens")
    print(f"Combined savings: {combined_savings}%")

    c = combined["cost_usd"]
    print(f"\nEstimated input cost (Claude Sonnet 4.6, ${DEFAULT_COST_PER_1M_TOKENS}/1M tokens):")
    print(f"  Naive: ${c['naive']:.4f}")
    print(f"  CRP:   ${c['crp']:.4f}")
    print(f"  Saved: ${c['savings']:.4f}")

    print("\nPer-skill breakdown:")
    for name, result in all_results.items():
        sr = result["session_5rounds"]
        print(f"  {name:15s} {sr['crp_total_tokens']:>8,} tokens ({sr['savings_percent']}% savings)")

    if report:
        out_path = Path("benchmark-report.json")
        report_data = {
            "combined_analysis": combined,
            "individual_skills": all_results,
            "method": list(all_results.values())[0]["method"] if all_results else "[estimated]",
        }
        out_path.write_text(json.dumps(report_data, indent=2, ensure_ascii=False) + "\n", encoding="utf-8")
        print(f"\nReport written to {out_path}")

    return 0


def main() -> int:
    parser = argparse.ArgumentParser(description="Token audit for CRP vs naive loading")
    parser.add_argument("--skill", default=None, help="Skill name (auto-detect if omitted)")
    parser.add_argument("--report", action="store_true", help="Write JSON to benchmark-report.json")
    args = parser.parse_args()

    return run_audit(args.skill, args.report)


if __name__ == "__main__":
    sys.exit(main())
